### ISSUE TITLE
Add support for Levoit Vital 100s air purifier

### ISSUE THREAD

**Issue by M1K3SD1** [webdjoe/pyvesync#181]
Could support for the new Vital 100s air purifier model be added please?

Model - LAP-V102S-WUS

**Comment by webdjoe** [comment]
I would need packet captures. Capturing packets has become more difficult due to ssl pinning, the only current way is by following the directions here in the feature requests section of the Readme

**Comment by M1K3SD1** [comment]
Working on getting the captures

**Comment by Purizze** [comment]
Hi, I want to help capture the packets for the 100s and already installed the Charles App.
How do I procceed from there?
What exactly am I supposed to capture?
Never did this before.

**Comment by M1K3SD1** [comment]
I'm doing this with an Android emulator and it's not that easy. I got the proxy working but I'm battling SSL pinning of the app currently. Will update.

**Comment by webdjoe** [comment]
My suggestion is look for as old a version of the app as you can possibly find the APK for that works with your device on the different site such as apkpure, apkmirror and aptoide. They are very aggressive in updating the SSL pinning methods.

**Comment by M1K3SD1** [comment]
Went back all the way to version 3.2.47. No luck...
Not sure what to from here to be honest.

Edit - this one isn't even compatible with my devices.

**Comment by webdjoe** [comment]
How did you bypass? Did you try the Frida universal bypass and the multi-unpinning script? 
https://codeshare.frida.re/@pcipolloni/universal-android-ssl-pinning-bypass-with-frida/
https://codeshare.frida.re/@akabe1/frida-multiple-unpinning/

**Comment by M1K3SD1** [comment]
No, I used Magisk and TrustMeAlready. I will take a look at this as well.
Thank you @webdjoe

**Comment by webdjoe** [comment]
Here's an article that might help https://petruknisme.medium.com/easy-way-to-bypass-ssl-pinning-with-objection-frida-beginner-friendly-58118be4094

The last resort would be to decompile and look for the API endpoints in the code but that is much more painful than using frida. Good Luck!

**Comment by M1K3SD1** [comment]
@webdjoe looks like I finally succeeded (that was pretty hard!). 
I have a capture that I saved as CHLSX (Charles XML).
How is it best to get you the file? (I prefer not to upload it here, although I redacted some of it).

Looks like this is the model: VS_WFON_APR_LAP-V102S-WUS_US

**Comment by webdjoe** [comment]
You can send it to my email (webdjoe at gmail). What did you use to capture?

**Comment by M1K3SD1** [comment]
I used Objection and captured with Charles proxy. The app was running in an Android emulator, but I had to use a different method of unpinning since the one mentioned in the guide you sent didn't work.

Will send in approx an hour.

**Comment by M1K3SD1** [comment]
@webdjoe file sent, let me know if this is good or if anything else is needed.
Thank you for the great work!

**Comment by webdjoe** [comment]
Check out #188 . The only methods I was able to incorporate were `turn_on`, `turn_off`, `change_fan_speed`, `auto_mode`, `manual_mode`. If you want to include turn display on and off, set child lock, timers and schedules please capture those packets. There is also a light detection attribute that I'm not sure what it does but let me know if that's something you think should be tracked in the instance.

FYI - the tests fail but that's on purpose until the tests are first run locally when there is a working product

**Comment by M1K3SD1** [comment]
Thank you @webdjoe, I will capture these as well.
About the light detection - this device has light detection which turns the device screen off when it's dark and turns it on when there's light. I can toggle the light in the room to capture this as well if it helps.

**Comment by M1K3SD1** [comment]
@webdjoe sent you a new capture with everything you mentioned.
Thanks!

**Comment by flatronic** [comment]
@M1K3SD1 Hey Mike, where did you buy your device? I own 3 from Germany and I guess they might have a slightly different name? (Model: ‎LAP-V102S-WEU) So if Modelname is important we might know the name as well :-)

**Comment by M1K3SD1** [comment]
@flatronic I'm in the US, so Amazon US :)
I guess the WUS vs WEU is the location of the model.

**Comment by M1K3SD1** [comment]
Hi @webdjoe, have you had the chance to look at the captures? Anything else I can help with?
Thank you for all the help!

**Comment by webdjoe** [comment]
@M1K3SD1 I incorporated the first bunch of packets you sent into #188  can you please test to make sure it is working correctly? The methods available should be `turn_on`, `turn_off`, `change_fan_speed`, `auto_mode`, `manual_mode`

@flatronic I will add LAP-V102S-WEU to #188, they should have the exact same functionality. Please test out the branch if you can

**Comment by M1K3SD1** [comment]
Thank you @webdjoe, honestly, I'm not sure how can I test it with my HA since I'm not running a Python env.
Is there a better way?

**Comment by sdrapha** [comment]
Best way to test it would be via CLI on any terminal. This is just the python library. Homeassistant runs a 'component', which then interfaces to this python library. It's a different piece of code, a whole extra layer to it. That HA component would also need a PR to support the device, AFTER the integration on the library here is done.

So, in a nutshell, start with any linux terminal, anywhere, VM, wsl, even Termux on android works.
Install python and python env, git clone the repo, switch to the right branch, activate the pyenv, install the library with pip3 install -e , start a interactive python session with python -i, then import the vesync module, and go from there with the commands.
There is more informations on the CONTRIBUTING.MD

**Comment by M1K3SD1** [comment]
@webdjoe looks like the device is not being added.

It looks like you added model "LAP-V102S-AAS" but mine is "LAP-V102S-WUS"?

```
2023-05-24 21:15:36,724 - DEBUG - =======call_api=============================
2023-05-24 21:15:36,724 - DEBUG - [post] calling '/cloud/v1/deviceManaged/devices' api
2023-05-24 21:15:36,724 - DEBUG - API call URL:
  https://smartapi.vesync.com/cloud/v1/deviceManaged/devices
2023-05-24 21:15:36,724 - DEBUG - API call headers:
  {"Content-Type": "application/json; charset=UTF-8", "User-Agent": "okhttp/3.12.1"}
2023-05-24 21:15:36,724 - DEBUG - API call json:
  {"timeZone": "America/Los_Angeles", "acceptLanguage": "en", "accountID": "##_REDACTED_##", "token": "##_REDACTED_##", "appVersion": "2.8.6", "phoneBrand": "SM N9005", "phoneOS": "Android", "traceId": "1684988136", "method": "devices", "pageNo": "1", "pageSize": "100"}
2023-05-24 21:15:37,135 - DEBUG - API response:

  {"traceId": "1684988136", "code": 0, "msg": "request success", "module": null, "stacktrace": null, "result": {"total": 2, "pageSize": 100, "pageNo": 1, "list": [{"deviceRegion": "US", "isOwner": true, "authKey": null, "deviceName": "Upstairs Air Purifier", "deviceImg": "https://image.vesync.com/defaultImages/deviceDefaultImages/vs_wfon_apr_lap_v102s.png", "cid": "##_REDACTED_##", "deviceStatus": null, "connectionStatus": null, "connectionType": "WiFi+BTOnboarding+BTNotify", "deviceType": "LAP-V102S-WUS", "type": "wifi-air", "uuid": "##_REDACTED_##", "configModule": "VS_WFON_APR_LAP-V102S-WUS_US", "macID": "##_REDACTED_##", "mode": null, "speed": null, "currentFirmVersion": null, "subDeviceNo": null, "subDeviceType": null, "deviceFirstSetupTime": "Mar 29, 2023 3:35:38 AM", "subDeviceList": null, "extension": null, "deviceProp": {"workMode": "auto", "AQLevel": 1, "BTMac": "60:55:f9:YY:YY:YY", "PM25": 5, "fanSpeedLevel": 255, "powerSwitch": 0, "wifiRssi": -45, "connectionStatus": "online", "wifiMac": "60:55:f9:XX:XX:XX"}}, {"deviceRegion": "US", "isOwner": true, "authKey": null, "deviceName": "Downstairs Air Purifier", "deviceImg": "https://image.vesync.com/defaultImages/deviceDefaultImages/vs_wfon_apr_lap_v102s.png", "cid": "##_REDACTED_##", "deviceStatus": null, "connectionStatus": null, "connectionType": "WiFi+BTOnboarding+BTNotify", "deviceType": "LAP-V102S-WUS", "type": "wifi-air", "uuid": "##_REDACTED_##", "configModule": "VS_WFON_APR_LAP-V102S-WUS_US", "macID": "##_REDACTED_##", "mode": null, "speed": null, "currentFirmVersion": null, "subDeviceNo": null, "subDeviceType": null, "deviceFirstSetupTime": "Mar 29, 2023 2:23:28 AM", "subDeviceList": null, "extension": null, "deviceProp": {"workMode": "auto", "AQLevel": 1, "BTMac": "60:55:f9:YY:YY:YY", "PM25": 5, "fanSpeedLevel": 255, "powerSwitch": 0, "wifiRssi": -42, "connectionStatus": "online", "wifiMac": "60:55:f9:XX:XX:XX"}}]}}

2023-05-24 21:15:37,135 - DEBUG - New device list initialized
2023-05-24 21:15:37,135 - DEBUG - Unknown device named Upstairs Air Purifier model LAP-V102S-WUS
2023-05-24 21:15:37,135 - DEBUG - Error - 'VeSync' object has no attribute 'unknown'
2023-05-24 21:15:37,136 - DEBUG - LAP-V102S-WUS device not added
2023-05-24 21:15:37,136 - DEBUG - Unknown device named Downstairs Air Purifier model LAP-V102S-WUS
2023-05-24 21:15:37,136 - DEBUG - Error - 'VeSync' object has no attribute 'unknown'
2023-05-24 21:15:37,136 - DEBUG - LAP-V102S-WUS device not added
2023-05-24 21:15:37,136 - DEBUG - Start updating the device details one by one
```

I tried changing to the correct model in vesyncfan.py but getting this: "No configuration found in purifier status"
Not sure why.

**Comment by webdjoe** [comment]
@M1K3SD1 Just pushed an update, included US and EU models

**Comment by M1K3SD1** [comment]
@webdjoe confirming that the modes you've mentioned above are all working now!
(turn_on, turn_off, change_fan_speed, auto_mode, manual_mode)

**Comment by M1K3SD1** [comment]
@webdjoe please commit the changes :)
Also, any plans to add the other missing methods I captured?

**Comment by webdjoe** [comment]
@M1K3SD1 I'm working on the new captures now, should have a working commit in a day or two.

**Comment by webdjoe** [comment]
@M1K3SD1 just pushed the update. Please let me know how it is working

**Comment by M1K3SD1** [comment]
Thank you @webdjoe,  it seems that there is a bit of confusion about some of the supported features with my device.
The basic features work fine, but I cannot control or get the light detection (it says no attribute 'light_detection_state'), and it says that display control is not implemented - although I do have both.

On the other hand, it let me set a night light although I do not have such a feature on the device (returned code -1).

**Comment by webdjoe** [comment]
@M1K3SD1 Did your local clone of #178 update? Those stub methods were replaced with the actual:

https://github.com/webdjoe/pyvesync/blob/4449693eee5df20295d0ec945e9af3c5864c7146/src/pyvesync/vesyncfan.py#L871

https://github.com/webdjoe/pyvesync/blob/4449693eee5df20295d0ec945e9af3c5864c7146/src/pyvesync/vesyncfan.py#L937

**Comment by M1K3SD1** [comment]
@webdjoe I just did again and verified, but I'm getting this:

`>>> fan.set_light_detection(on)
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
AttributeError: 'VeSyncVital100S' object has no attribute 'set_light_detection'`

Unless I'm doing something completely wrong.

**Comment by webdjoe** [comment]
Nope, I renamed the module without renaming in the configuration dictionary. That's what I get for not building tests.

**Comment by M1K3SD1** [comment]
Hah, no worries, I'll wait for a fix.

**Comment by den-mac** [comment]
Sorry, stumbled upon this thread while looking for a solution for my Vital 100S air purifier, which the current VeSync integration isn't finding.

Maybe I need to wait - I'm unclear how all this works. I am guessing:

- some testing of https://github.com/webdjoe/pyvesync/pull/188 still needs to be done, as in fix the various py and other tests
- at some point it will get integrated into the live code and an integration update will get pushed out?

Do I assume correctly here? Is there some way I can test it since some folks say that some of it works?

Thanks! and sorry for the noob questions here...

**Comment by webdjoe** [comment]
I've done some debugging and the class is changed to `VeSyncVital`, along with the confi dictionary.  Try to delete the branch completely and clone again. If it tries to instantiate the `VeSyncVital100S` class it should throw an error on the initial `get_devices` call, not when trying to invoke a method because `VeSyncVital100S` doesn't exist.

The config dictionary has the changes:
![image](https://github.com/webdjoe/pyvesync/assets/8164486/77fc7c12-3fab-4afc-bd6e-eb0035bb00be)


And the class name has been changed:
![image](https://github.com/webdjoe/pyvesync/assets/8164486/84bc62b1-e917-46ac-8e74-322ee6513a59)

**Comment by M1K3SD1** [comment]
Hi @webdjoe, I just cloned again and the command seems to be working this time. I apologize if I screwed it up last time :)

I noticed that there are some features that are there but not mentioned in the readme, or maybe you haven't added them yet.
For example, the child lock and schedule/timer?

**Comment by M1K3SD1** [comment]
Hi @webdjoe, sorry for pinging, any chance you can merge it soon, please?

**Comment by webdjoe** [comment]
Just got back from vacation, I will get this released this weekend!

**Comment by M1K3SD1** [comment]
Hi @webdjoe, I imagine you're busy but can you please merge the changes? I'd love to be able to control at least some of the settings.

Much appreciated!

**Comment by webdjoe** [comment]
Sorry for the delay, released in 2.18 on PyPi!

**Comment by webdjoe** [comment]
Closing this and leaving #178 open for any bugs
